### PR TITLE
Make

### PR DESCRIPTION
**Pull request by danking**
### Summary of Changes

- make the vectorized IBS tests depend on libsmidpp, I have no idea how this wasn't failing builds before.

- fix `generate-build-info.sh` to actually set `BRANCH` instead of setting the `DATE` to the current git branch. Also remove unnecessary parameter passing to `echo_build_properties`.

- remove `doctest` as a separate build step (it's now part of python tests)

- fix `DOCS_STATUS` in `hail-ci-build.sh`

- remove some unused files: `list_pypi_versions.py`, `publish-to-pypi.sh` (which duplicates, wrongly, `python/deploy.sh`

- simplify `python/deploy.sh`, by managing copied files in make, we no longer clean up files, but the JAR is only 30 MB anyway.

- small change to `conftest.py` triggered a bad diff: I added a `try`/`finally` block to ensure we always return to the original working directory. I also fixed the path.

- Use pytest [recommended directory structure](https://docs.pytest.org/en/latest/goodpractices.html) when your tests directory is a python module [1]. In particular, we now use:

```
python/
- setup.py
- src/
  - hail/
    - __init__.py
    - ...
- tests/
  - __init__.py
  - ...
```

- Calculate number of cores using python's multiprocessing and use that as a default PARALLELISM parameter

- Move non-java/scala specific functionality out of `build.gradle` and into a `Makefile`.

  - The resulting rules are more succinct and correctly rely on file-system modification dependencies.

  - No use of `SPARK_HOME` and `PYTHONPATH`, and limited use of `PYSPARK_SUBMIT_ARGS`. Python tests now rely on the python package directly which handles correctly handles dependencies like `pyspark`.

  - There are also some phony targets for convenience: `jar`, `zip`, `pip-install`, `docs`, and `docs-no-test`.

- Fix configuration of Spark version for the python package. The version is written by make into `python/spark_version` and read by `python/setup.py`. Many of the tests pass against 2.3.0, but there's some floating point value changes.

- add breezeVersions for all currently released Spark versions greater than 2.2.0.

- For developers, require python package `py` version 1.7.0 or later to allow `pytest` to test an installed package while loading the doctest expressions from the source code. (We could also determine where hail was installed and pass that path to pytest instead of `python/src`, but using the environment variable `PY_IGNORE_IMPORTMISMATCH` seems simple and safe enough)

---

#### env_var.mk

This is a Makefile that is intended to be `include`d by other Makefiles. It defines a [multi-line variable](https://www.gnu.org/software/make/manual/html_node/Multi_002dLine.html) that [takes arguments](https://www.gnu.org/software/make/manual/html_node/Call-Function.html#Call-Function) (known in any reasonable language as a "function").

It is intended to be used like this:

```
VERSION = 30
$(eval $(call ENV_VAR,VERSION))

build: env/VERSION
build:
  ... $(VERSION) ...
```

Each time this Makefile is executed, at Makefile parse-time, `make` evaluates the `ifneq` to compare the current value of the variable to the previously used value (if any). If they differ, a phony (ergo always needs to be rebuilt) target is dynamically generated. That target will force a execution of any dependent targets, in the example above, it will force `build` to be executed. If the variable's current value and it's previous value do not differ, no target is generated and thus nothing is executed.

#### MAKEFLAGS += --no-builtin-rules

This disables all the automatic rules, making Makefile parsing/interpretation faster.

#### .SUFFIXES:

This sets the suffixes to nothing, disabling a bunch of implicit rules. For example, the rule for compiling C files.

#### Breeze versions

I found the breeze versions by manually looking at the Spark pom file at the tags for each release, starting with [2.2.0](https://github.com/apache/spark/blob/v2.2.0/pom.xml#L654).

---

[1] A python module is any folder containing a file called `__init__.py`

### CODE REVIEW

**Comment by tpoterba** [comment]
> I got frustrated with ./gradlew testPython because it requires SPARK_HOME to be set

Can't we just require pyspark 2.2 to be pip-installed and remove that stuff? Should work.

I need a bit of time to digest. I was under the impression we were using a pytest recommended layout, and going back to the doc you linked, we have the recommended structure under "Tests outside application code":

```
setup.py
mypkg/
    __init__.py
    app.py
    view.py
tests/
    test_app.py
    test_view.py
    ...
```

It looks like the `src/` structure is recommended in the following case:

> ...  This is problematic if you are using a tool like tox to test your package in a virtual environment, because you want to test the installed version of your package, not the local code from the repository.

> In this situation, it is strongly suggested to use a src layout where application root package resides in a sub-directory of your root:

Digging a little bit more, the pytest page references [this blog](https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure) for reasons to use `src/`. But some of the comments indicate that `pip install -e` won't work with this structure, since the symlink it creates will be incorrect. It's possible this has been fixed since the issues were posted in 2012.

I like the arguments there about testing the installed package so that the deploy is also tested.

My reluctance mostly comes from knowing that I spend an hour trying to reconfigure intelliJ every time our directory structure changes.

Let's discuss today. As above, I think this is fine but want to think over it a bit more.

**Comment by tpoterba** [comment]
having thought a bit more, I think this is a very good idea.

**Comment by tpoterba** [comment]
I do, however, think we should hold off until the several spicy meatballs are cooked and committed.

**Comment by tpoterba** [comment]
ah, I see. I'm not sure why we added that __init__.py in the first place... it's empty

**Comment by danking** [comment]
Yeah you're right about `pyspark` though, if we `pip install pyspark` then we needn't edit PYTHONPATH or set `SPARK_HOME` afaik.

**Comment by danking** [comment]
I suspect someone added it after extensively wrestling with pytest because without it, pytest would fail on `import hail as hl` because the path search would stop too soon.

**Comment by danking** [comment]
@tpoterba I updated the main message with a better explainer on the changes.

File: hail/python/tests/helpers.py [hunk 1]
```diff
@@ -41,18 +40,8 @@ def resource(filename):
...
-        while not os.path.exists(os.path.join(path, 'build.gradle')):
-            path = os.path.join(path, '..')
-            i += 1
-            if i > 100:
-                raise EnvironmentError("Hail tests must be run from inside the Hail git repository")
-        _doctest_dir = os.path.join(path, 'python', 'hail', 'docs', 'data')
-
-    return os.path.join(_doctest_dir, filename)
+    import pkg_resources
+    return pkg_resources.resource_filename('hail', 'docs/data/'+filename)
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/python/tests/helpers.py]: You'd think, but [pkg_resources says don't do that](https://stackoverflow.com/a/20885799). 🤷‍♀️

**Comment by danking** [comment]
```
(hail) dking@wmb16-359 # time make jar
make: Nothing to be done for `jar'.
make jar  0.04s user 0.06s system 88% cpu 0.115 total
```

File: hail/python/tests/helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/python/tests/helpers.py]: wow

**Comment by danking** [comment]
moved to https://github.com/hail-is/hail/pull/5194

**Pull request closed without merging**